include/oneapi/dpl/internal/async_impl/async_impl_hetero.h

```diff
@@ -0,0 +1,202 @@
+/*
```

Comment:
What's a goal of the `async_impl_hetero.h ` file ?
Why do we need to add a special "async" hetero algorithm patterns level?
In my vision, we can re-use the existing hetero algorithm patterns level.

Reply:
Not sure if it can be done. For example existing pattern with `void` return type have to be changed to `future<void>`, can result at least in a warning at compile time. If we add code branches for sync and async we have to modify more than just the hetero layer from my perspective.

Reply:
In my vison, we can re-use the existing hetero algorithm patterns level.
A caller (on on the PSTL highest level), after call a algorithm pattern, should call 'wait' and return result for sycn version and return a future for async version, nothing more.

Reply:
I like that idea and think that is the right way to go. However, it would require extensive changes to the existing layout. Which I was trying to avoid with the initial subset of async algorithm.

Reply:
I think, the extensive changes to the current layout are acceptable for a good design :)
BTW, "extensive" , but not big and complicated.

Reply:
I have created an issue: https://github.com/oneapi-src/oneDPL/issues/135

Reply:
That looks strange (as in a case of SYLC blackened) - to duplicate the same backend code, merge it, and after - remove one.
I can do it (keep the same hetero ago patterns ) in 1-2 days (on your branch).